pkg/controller/populators/forklift-populator.go

```diff
@@ -0,0 +1,584 @@
+package populators
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+
+	"fmt"
+	"regexp"
+	"strconv"
+	"time"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/ptr"
+	"kubevirt.io/containerized-data-importer-api/pkg/apis/forklift/v1beta1"
+	cc "kubevirt.io/containerized-data-importer/pkg/controller/common"
+	featuregates "kubevirt.io/containerized-data-importer/pkg/feature-gates"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	forkliftPopulatorName  = "forklift-populator-controller"
+	populatorContainerName = "populate"
+	populatorPodPrefix     = "populate"
+	populatorPodVolumeName = "target"
+	mountPath              = "/mnt/"
+	devicePath             = "/dev/block"
+)
+
+const apiGroup = "forklift.kubevirt.io"
+
+var (
+	supportedPopulators = map[string]client.Object{
+		"OvirtVolumePopulator":     &v1beta1.OvirtVolumePopulator{},
+		"OpenstackVolumePopulator": &v1beta1.OpenstackVolumePopulator{},
+	}
+)
+
+var errCrNotFound = errors.New("populator CR not found")
+
+// ForkliftPopulatorReconciler members
+type ForkliftPopulatorReconciler struct {
+	ReconcilerBase
+	importerImage       string
+	ovirtPopulatorImage string
+}
+
+// NewForkliftPopulator creates a new instance of the forklift controller.
+func NewForkliftPopulator(
```

Comment:
Do we want to add a few e2e assertions that the populator gets used instead of legacy?

Reply:
So these populators are not replacing legacy DV flows, rather, they're new functionality that did not exist in CDI before?
If that's true, I guess this is what makes it tricky to test here?

Reply:
yes, you could say the ovirt populator can replace the imageio data source, but there is no feature parity. The populator does not support checkpoints like the DV datasource for example, so it will still be used for warm migration from ovirt

I would be interested in having tests for this other than unit tests, but given that it doesn't seem trivial to do and we need some buffer to test this with forklift, I'd do it separately